server/chat-plugins/help.ts

```diff
@@ -0,0 +1,473 @@
+				if (raw) return string.trim();
+				// escape otherwise
+				return string.replace(/[\\^$.*+?()[\]{}]/g, '\\$&').trim();
+			});
+			return split.map(term => {
+				if (term.startsWith('!')) {
+					return `^(?!.*${term.slice(1)})`;
+				}
+				if (!term.trim()) return null;
+				return `(?=.*?(${term.trim()}))`;
+			}).filter(Boolean).join('');
+		}).filter(Boolean).join('');
+	}
+	validateRegex(regex: string) {
+		try {
+			new RegExp(regex).test('');
+		} catch (e) {
+			if (e.message.includes("regular expression")) throw new Chat.ErrorMessage(e.message);
+			throw e;
+		}
+		return true;
+	}
+	match(question: string, faq: string) {
+		const regexes = this.data.pairs[faq];
+		if (!regexes) return;
+		for (const regex of regexes) {
+			const regexString = new RegExp(regex, "i");
+			if (regexString.test(Chat.normalize(question))) return {faq, regex};
+		}
+		return;
+	}
+	log(entry: string, faq: string, expression: string) {
+		if (!this.data.stats) this.data.stats = {};
+		const day = Chat.toTimestamp(new Date).split(' ')[0];
+		if (!this.data.stats[day]) this.data.stats[day] = {};
+		const today = this.data.stats[day];
+		const log: LoggedMessage = {
+			message: entry,
+			faqName: faq,
+			regex: expression,
+		};
+		const stats = {
+			matches: today.matches || [],
+			total: today.matches ? today.matches.length : 0,
+		};
+		const dayLog = Object.assign(this.data.stats[day], stats);
+		dayLog.matches.push(log);
+		dayLog.total++;
+		return this.writeState();
+	}
+	writeState() {
+		return FS(PATH).writeUpdate(() => JSON.stringify(this.data));
+	}
+	loadFaqs() {
+		this.roomFaqs = JSON.parse(FS(ROOMFAQ_FILE).readIfExistsSync() || `{"help":{}}`).help;
+		for (const key in this.data.pairs) {
+			if (!this.roomFaqs[key]) delete this.data.pairs[key];
+		}
+		this.writeState();
+		return this.roomFaqs;
+	}
+	tryAddRegex(inputString: string, raw?: boolean) {
+		let [args, faq] = inputString.split('=>');
+		faq = this.getFaqID(toID(faq)) as string;
+		if (!faq) throw new Chat.ErrorMessage("Invalid FAQ.");
+		if (!this.data.pairs) this.data.pairs = {};
+		if (!this.data.pairs[faq]) this.data.pairs[faq] = [];
+		const regex = raw ? args.trim() : this.stringRegex(args, raw);
+		if (this.data.pairs[faq].includes(regex)) {
+			throw new Chat.ErrorMessage(`That regex is already stored.`);
+		}
+		this.validateRegex(regex);
+		this.data.pairs[faq].push(regex);
+		return this.writeState();
+	}
+	tryRemoveRegex(faq: string, index: number) {
+		faq = this.getFaqID(faq) as string;
+		if (!faq) throw new Chat.ErrorMessage("Invalid FAQ.");
+		if (!this.data.pairs) this.data.pairs = {};
+		if (!this.data.pairs[faq]) throw new Chat.ErrorMessage(`There are no regexes for ${faq}.`);
+		if (!this.data.pairs[faq][index]) throw new Chat.ErrorMessage("Your provided index is invalid.");
+		this.data.pairs[faq].splice(index, 1);
+		this.writeState();
+		return true;
+	}
+}
+
+export const Answerer = new HelpAnswerer(helpData);
```

Comment:
Why is this exported?

Reply:
Eval purposes, mostly, didn't remember to remove it.

Reply:
I don't think it's wrong for things like this to be exported. It is, in fact, useful for monkeypatching.

Reply:
Exports define your API. It doesnt make sense to me that this would be part of an API? Furthermore, why are you monkeypatching chat plugins when theyre already hotpatchable, am I missing something?